src/tests/ftest/util/job_manager_utils.py

```diff
@@ -439,3 +460,507 @@ def assign_environment_default(self, env_vars):
+                            self.log.info("      %s", line)
+                    else:
+                        self.log.info(
+                            "    %s: rc=%s, output: %s",
+                            node_set, code, output_buffer)
+
+        # Report any errors through an exception
+        if not status:
+            raise CommandFailure(
+                "Error(s) detected gathering {} log data on {}".format(
+                    self._systemctl.service.value, NodeSet.fromlist(hosts)))
+
+        # Return the successful command output per set of hosts
+        return log_data
+
+    def display_log_data(self, log_data):
+        """Display the journalctl log data.
+
+        Args:
+            log_data (dict): dictionary of journalctl log output.
+        """
+        self.log.info("Journalctl output:")
+        for line in self.str_log_data(log_data).split("\n"):
+            self.log.info(line)
+
+    @staticmethod
+    def str_log_data(log_data):
+        """Get the journalctl log data as a string.
+
+        Args:
+            log_data (dict): dictionary of journalctl log output.
+
+        Returns:
+            str: the journalctl log data
+
+        """
+        data = []
+        for node_set in sorted(log_data):
+            data.append("  {}:".format(node_set))
+            for line in log_data[node_set]:
+                data.append("    {}".format(line))
+        return "\n".join(data)
+
+    def check_logs(self, pattern, since, until, quantity=1, timeout=60):
+        """Check the command logs on each host for a specified string.
+
+        Args:
+            pattern (str): regular expression to search for in the logs
+            since (str): search log entries from this date.
+            until (str, optional): search log entries up to this date. Defaults
+                to None, in which case it is not utilized.
+            quantity (int, optional): number of times to expect the search
+                pattern per host. Defaults to 1.
+            timeout (int, optional): maximum number of seconds to wait to detect
+                the specified pattern. Defaults to 60.
+
+        Returns:
+            bool: whether or not the search string was found in the logs on each
+                host
+
+        """
+        self.log.info(
+            "Searching for '%s' in '%s' output on %s",
+            pattern, self._systemctl, self._hosts)
+
+        log_data = None
+        detected = 0
+        complete = False
+        timed_out = False
+        start = time.time()
+
+        # Search for patterns in the subprocess output until:
+        #   - the expected number of pattern matches are detected (success)
+        #   - the time out is reached (failure)
+        #   - the service is no longer running (failure)
+        while not complete and not timed_out and self.service_running():
+            detected = 0
+            log_data = self.get_log_data(self._hosts, since, until, timeout)
```

Comment:
Hrm.  This seems to result in a spew of:

```
2021-02-01 07:42:37,239 job_manager_util L0917 INFO | Searching for '(SCM format required)(?!;)' in 'sudo -n systemctl start daos_server.service' output on wolf-134
2021-02-01 07:42:37,416 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:37,417 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:37,768 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:37,769 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:38,117 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:38,117 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:38,469 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:38,469 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:38,820 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:38,820 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:39,179 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:39,179 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:39,536 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:39,536 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:39,891 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:39,891 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:40,244 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:40,244 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:40,612 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:40,612 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:40,964 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:40,964 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:41,316 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:41,316 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:41,670 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:41,671 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:42,023 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:42,023 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:42,377 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:42,377 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:42,730 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:42,730 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:43,086 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:43,086 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:43,443 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:43,444 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:43,801 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:43,801 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:44,157 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:44,157 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:44,512 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:44,512 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:44,873 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:44,873 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:45,233 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:45,233 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:45,593 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:45,593 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:45,954 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:45,954 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:46,311 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:46,311 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:46,667 job_manager_util L0768 INFO |   Detected daos_server.service states: active=wolf-134
2021-02-01 07:42:46,667 job_manager_util L0810 INFO | Gathering log data on wolf-134: sudo journalctl --unit=daos_server.service --since="2021-02-01 07:42:37"
2021-02-01 07:42:46,849 job_manager_util L0964 INFO |  subprocess startup detected - 1/1 '(SCM format required)(?!;)' messages detected in 9.60953402519/30 seconds
```

While waiting for the servers to start.

The journal can be *tail*ed like `tail -f` does with a `-f` option.  It would be more ideal to start a single process to tail, waiting for the desired output, or a timeout, whichever comes first, and not return until one of those has been achieved.  Then it's a single remote call, without a few dozen lines of output and a few dozen remote calls.

Reply:
I agree an improvement could be made in this space.  My question with the tail approach is how do we perform the search? Do we tail into a file as a background process so we can grep the file in a loop and then kill the background process?  

I could see an option where we clush a set of bash commands that loops running the current journalctl command (no tail) and then greps for the intended string - exiting the loop if all matches are found.  The loop timeout should be handled in the bash commands such that it can mimic the current behavior of only including the output of the last journalctl message.  Basically setting up the command to clush that looks something like:
```
command= (
    "if timeout 30 bash -c ' "
    "    pattern=\"''(SCM format required)''\"; "
    "    pattern_qty=2; "
    "    detected=0; "
    "    active=\"active\"; "
    "    while [[ ${detected} != ${pattern_qty} ]] && [[ ${active} -eq \"active\" ]]; do "
    "        active=$(systemctl is-active daos_server.service); "
    "        detected=$(sudo journalctl --unit=daos_server.service --since=\"${date}\" | grep -Ec \"${pattern}\"); "
    "        echo \"  ${detected}/${pattern_qty} ${pattern} messages detected; ${service} is ${active}\"; "
    "    done'; "
    "then "
    "    exit 0; "
    "else "
    "    echo \"  TIMEOUT detected - ${detected}/${pattern_qty} ${pattern} messages detected in 30 seconds\"; "
    "    sudo journalctl --unit=daos_server.service --since=\"${date}\"; "
    "    exit 1; "
    "fi")
```

I wonder how we expect users to perform this action with a large number of servers.  Would a dmg command that could inform the user when all the servers are ready for format and when all the I/O servers are running.  The later can probably be assessed through `dmg system query` returning "Joined" for all ranks, but it would still require looping the command.

Can/Should this be addressed as an a separate PR enhancement?  Or should it be changed now?

Reply:
> I agree an improvement could be made in this space. My question with the tail approach is how do we perform the search? Do we tail into a file as a background process so we can grep the file in a loop and then kill the background process?

I usually try to avoid `grep` and `tail` into a `while read -t $TIMEOUT` loop, looking for my expectation, breaking out of the loop when I have found it, or the overall waiting time is done.  If you really need a regex match (and want to reach for `grep`) you can use bash's built in `[[ $line =~ $REGEX ]]`.

> I wonder how we expect users to perform this action with a large number of servers.

Indeed, I have pondered that a number of times and even asked but not really gotten a suitable answer.  It's probably worthy of a ticket.  If it's even a small problem for us at our small scale, the problem is only going scale with the cluster.  It does seem to be a more-than-insignificant usability issue.

> Would a dmg command that could inform the user when all the servers are ready for format and when all the I/O servers are running. The later can probably be assessed through `dmg system query` returning "Joined" for all ranks, but it would still require looping the command.

Or (optionally) blocking, with a timeout perhaps.

> Can/Should this be addressed as an a separate PR enhancement? Or should it be changed now?

I guess I'd leave that up to your discretion based on your expected level of effort (and delay) to implement.  And I suppose your confidence on getting it right the first time, to avoid flailing on it.  Although with `Quick-Functional: true` you could even get through a few iterations pretty quickly.